pkg/activator/handler/concurrency_reporter.go

```diff
@@ -86,81 +87,77 @@ func (cr *ConcurrencyReporter) Run(stopCh <-chan struct{}) {
 
 	for {
 		select {
 		case event := <-cr.reqCh:
-			switch event.Type {
-			case network.ReqIn:
-				incomingRequestsPerKey[event.Key]++
-
-				// Report the first request for a key immediately.
-				if _, ok := outstandingRequestsPerKey[event.Key]; !ok {
-					reportedFirstRequest[event.Key] = 1.
+			stat := stats[event.Key]
+			if stat == nil {
+				stat = network.NewRequestStats(event.Time)
+				stats[event.Key] = stat
+
+				// Only generate a from 0 event if this is an incoming request.
+				if event.Type == network.ReqIn {
+					reportedFirstRequest[event.Key] = 1
 					cr.statCh <- []asmetrics.StatMessage{{
 						Key: event.Key,
 						Stat: asmetrics.Stat{
-							// Stat time is unset by design. The receiver will set the time.
+							// Stat time is unset by design. Will be set by receiver.
 							PodName:                   cr.podName,
 							AverageConcurrentRequests: 1,
 							// The way the check above is written, this cannot ever be
-							// anything else but 1.
-							// In theory consider the situation where within the same
-							// reporting period a request arrived and terminated
-							// (making outstandingRequestsPerKey for that key 0) and
-							// then the same situation happened again, and again...
-							// Thus this might be > 1, but we only check for `!ok` not
-							// `!ok || val==0`, which means this is not reported until
-							// the reporting period channel ticks.
-							// Thus this will always be 1.
+							// anything else but 1. The stats map key is only deleted
+							// after a reporting period, so we see this code path at most
+							// once per period.
 							RequestCount: 1,
 						},
 					}}
 				}
-				outstandingRequestsPerKey[event.Key]++
-			case network.ReqOut:
-				outstandingRequestsPerKey[event.Key]--
 			}
-		case <-reportCh:
-			messages := make([]asmetrics.StatMessage, 0, len(outstandingRequestsPerKey))
-			for key, concurrency := range outstandingRequestsPerKey {
+
+			stat.HandleEvent(event)
+		case now := <-reportCh:
+			messages := make([]asmetrics.StatMessage, 0, len(stats))
+			for key, stat := range stats {
+				report := stat.Report(now)
 				firstAdj := reportedFirstRequest[key]
-				averageConcurrentRequests := concurrency - firstAdj
-				requestCount := incomingRequestsPerKey[key] - firstAdj
 
-				if concurrency == 0 {
-					delete(outstandingRequestsPerKey, key)
-					averageConcurrentRequests = 0
+				// This is only 0 if we have seen no activity for the entire reporting
+				// period at all.
+				if report.AverageConcurrency == 0 {
+					delete(stats, key)
 				}
 
+				// Subtract the request we already reported when first seeing the
+				// revision. We report a min of 0 here because the initial report is
+				// always a concurrency of 1 and the actual concurrency reported over
+				// the reporting period might be < 1.
+				averageConcurrency := math.Max(report.AverageConcurrency-firstAdj, 0)
```

Comment:
potentially a bit confusing that for this to be called `averageConcurrency` when `report.AverageConcurrency` is still live later in the block. It wasn't clear on first reading, to me at least, why the `averageConcurrency` here (the one that's at least 0 and adjusted by first report) is the correct one to use as `isometrics.Stat.AverageConcurrentRequests` a few lines later, but the original `report.AverageConcurrency` is the one to use in `cr.reportToMetricsBackend` a couple lines after that.

Maybe a rename to something like `adjustedAverageConcurrency` to avoid any confusion?

Reply:
Good point 👍 done!